File: implement-shell-tools/cat/cat.js
import { program } from "commander";
import { promises as fs } from "node:fs";

program
  .name("check-for-cat")
  .description("Implement my own version of cat")
  .argument("<paths...>", "The file paths to process")
  .option("-n", "Number lines")
  .option("-b", "Number non-blank lines");

program.parse();

const paths = program.args;
const showN = program.opts().n;
const showB = program.opts().b;

let lineNumber = 1;

for (const path of paths) {
  const content = await fs.readFile(path, "utf-8");

  const endsWithNewline = content.endsWith("\n");
  let lines = content.split("\n");
  if (endsWithNewline) {
    lines.pop();
  }

  lines = lines.map((line) => {
    if (showB) {
   
      if (line === "") {
        return line;
      }
      const numbered = String(lineNumber).padStart(6, " ") + "\t" + line;
      lineNumber++;
      return numbered;
    } else if (showN) {
    
      const numbered = String(lineNumber).padStart(6, " ") + "\t" + line;
      lineNumber++;
      return numbered;
    } else {
     
      return line;
    }
  });

  let output = lines.join("\n");
  if (endsWithNewline) {
    output += "\n";
  }
  process.stdout.write(output);
}

File: implement-shell-tools/cat/cat.py
import argparse

parser = argparse.ArgumentParser(
    prog="check-for-cat",
    description="Implement my own version of cat in python",
)

parser.add_argument("paths", nargs="+", help="The file paths to process")
parser.add_argument("-n", action="store_true", help="Number lines")
parser.add_argument("-b", action="store_true", help="Number non-blank lines")

args = parser.parse_args()

paths = args.paths
show_n = args.n
show_b = args.b

line_number = 1

for path in paths:
    with open(path, "r", encoding="utf-8") as f:
        content = f.read()

    ends_with_newline = content.endswith("\n")
    lines = content.split("\n")
    if ends_with_newline:
        lines.pop()

    new_lines = []
    for line in lines:
        if show_b:
            if line == "":
                new_lines.append(line)
            else:
                numbered = str(line_number).rjust(6, " ") + "\t" + line
                line_number += 1
                new_lines.append(numbered)
        elif show_n:
            numbered = str(line_number).rjust(6, " ") + "\t" + line
            line_number += 1
            new_lines.append(numbered)
        else:
            new_lines.append(line)
    lines = new_lines

    output = "\n".join(lines)
    if ends_with_newline:
        output += "\n"
    print(output, end="")